server/svix-server/src/core/permissions.rs

```diff
@@ -101,6 +101,44 @@ where
     }
 }
 
+pub struct ApplicationWithMetadata {
+    pub app: application::Model,
+    pub metadata: Option<applicationmetadata::Model>,
+}
+
+#[async_trait]
+impl<B> FromRequest<B> for ApplicationWithMetadata
+where
+    B: Send,
+{
+    type Rejection = Error;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self> {
+        let permissions = permissions_from_bearer(req).await?;
+
+        let Path(ApplicationPathParams { app_id }) =
+            ctx!(Path::<ApplicationPathParams>::from_request(req).await)?;
+        let Extension(ref db) = ctx!(Extension::<DatabaseConnection>::from_request(req).await)?;
+        let (app, metadata) = ctx!(
+            application::Model::get_with_application_metadata(
+                db,
+                permissions.org_id(),
+                app_id.to_owned()
+            )
+            .await
+        )?
+        .ok_or_else(|| HttpError::not_found(None, None))?;
+
+        if let Some(permitted_app_id) = permissions.app_id() {
```

Comment:
What's scary to me is that this logic is now duplicated, so it's another place to update this logic once we change something.